cli/bench/lsp.rs

```diff
@@ -0,0 +1,274 @@
+}
+
+#[derive(Debug, Deserialize)]
+struct LspResponse {
+  pub result: Option<Value>,
+  pub id: u32,
+  pub error: Option<LspResponseError>,
+}
+
+#[derive(Debug, Deserialize)]
+struct LspResponseError {
+  pub code: i32,
+  pub message: String,
+  pub data: Option<Value>,
+}
+
+#[derive(Debug, Deserialize)]
+struct LspNotification {
+  pub method: String,
+  pub params: Option<Value>,
+}
+
+fn create_basic_runtime() -> runtime::Runtime {
+  runtime::Builder::new_current_thread()
+    .enable_io()
+    .enable_time()
+    // for performance purposes, we might as well allow this to go to as
+    // many threads as it can, so that the harness doesn't become the bottle
+    // neck
+    // .max_blocking_threads(32)
+    .build()
+    .unwrap()
+}
+
+pub(crate) fn run_local<F, R>(future: F) -> R
+where
+  F: std::future::Future<Output = R>,
+{
+  let rt = create_basic_runtime();
+  rt.block_on(future)
+}
+
+struct LspClient {
+  pub output: Lines<BufReader<ChildStderr>>,
+  reader: BufReader<ChildStdout>,
+  start: Instant,
+  writer: BufWriter<ChildStdin>,
+}
+
+impl LspClient {
+  pub fn new(deno_exe: &PathBuf) -> Result<Self, AnyError> {
+    let mut child = Command::new(deno_exe)
+      .kill_on_drop(true)
+      .args(&["lsp"])
+      .stderr(Stdio::piped())
+      .stdin(Stdio::piped())
+      .stdout(Stdio::piped())
+      .spawn()?;
+
+    let stdout = child
+      .stdout
+      .take()
+      .ok_or_else(|| generic_error("could not get handle to stdout"))?;
+    let reader = BufReader::new(stdout);
+
+    let stdin = child
+      .stdin
+      .take()
+      .ok_or_else(|| generic_error("could not get handle to stdin"))?;
+    let writer = BufWriter::new(stdin);
+
+    let stderr = child
+      .stderr
+      .take()
+      .ok_or_else(|| generic_error("could not get handle to stderr"))?;
+    let output = BufReader::new(stderr).lines();
+
+    tokio::spawn(async move {
+      child.wait().await.expect("could not exit properly");
+    });
+
+    Ok(Self {
+      output,
+      reader,
+      start: Instant::now(),
+      writer,
+    })
+  }
+
+  pub fn duration(&self) -> u128 {
+    self.start.elapsed().as_millis()
+  }
+
+  pub async fn read<V>(&mut self) -> Result<V, AnyError>
+  where
+    V: de::DeserializeOwned,
+  {
+    let mut content_length = 0_usize;
+    loop {
+      let mut buf = String::new();
+      self.reader.read_line(&mut buf).await?;
+      if let Some(captures) = CONTENT_TYPE_REG.captures(&buf) {
+        let content_length_match = captures
+          .get(1)
+          .ok_or_else(|| generic_error("missing capture"))?;
+        content_length = content_length_match.as_str().parse::<usize>()?;
+      }
+      if &buf == "\r\n" {
+        break;
+      }
+    }
+
+    let mut msg_buf = vec![0_u8; content_length];
+    let bytes_read = self.reader.read(&mut msg_buf).await?;
+    assert_eq!(bytes_read, content_length);
+    let msg = serde_json::from_slice(&msg_buf)?;
+
+    Ok(msg)
```

Comment:
Nice mini-parser.

This could be made into a standalone function that operates on an abstract AsyncRead. Then we could pass it some canned data to see if it parses it correctly.

```rust
fn parse_lsp_message<R>(reader: R) -> Result<Vec<u8>, AnyError>
 where
    R: AsyncRead
{
    /// ... 
}

fn test_parser() {
   let msg = "content-length: 11\r\n\r\nhello world";
   let buf = std::io::Cursor::new(msg.as_bytes());
   assert_eq!(parse_lsp_message(buf), "hello world".as_bytes());
}

 
```